### PR TITLE
Feature/brinkman

### PR DESCRIPTION
**Pull request by timfelle**
Introduction of a Brinkman forcing term.

The idea of this is to enforce regions of low permeability to block flow in certain regions.
Included is an example with a Stanford bunny placed in a square duct as a block of resistance.

Still some todo points, but wanted to open it up so I could start to get some feedback

- [x] Documentation of the example and the source term in general, including an image or two.
- [ ] Device acceleration for computation of the brinkman term.
- [ ] Add open mesh example.
- [ ] Trim aabb_tree, triangles outside the fluid domain. -> Postponed, there are some situations which require this to be done with care.
- [ ] Expand to utilize point zones. -> moved to sepparate PR

### CODE REVIEW

**Comment by timfelle** [comment]
Do you have any preference on code structure? 

- I currently have implementation details in the source_terms/brinkman folder but i can move it to source_terms/bcknd if that makes it clearer.
- The AABB tree and could be moved to the mesh folder where the Octree also lives.

**Comment by njansson** [comment]
Is this ready for review now?

**Comment by timfelle** [comment]
Lets get it out there. Ill do seperate PR's for the other extensions to it as they are ready. This way others can get to experiment with it.

**Comment by timfelle** [comment]
Ups, wrong branch

File: src/source_terms/bcknd/cpu/brinkman_source_term_cpu.f90 [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  use math, only: subcol3
+  use field_registry, only: neko_field_registry
+  implicit none
+  private
+
+  public :: brinkman_source_term_compute_cpu
+
+contains
+
+  !> Computs the constant source term on the cpu.
```
Comment (timofeymukha) [inline src/source_terms/bcknd/cpu/brinkman_source_term_cpu.f90:47]:
```suggestion
  !> Computes the Brinkman source term on the cpu.
```

File: src/source_terms/bcknd/device/brinkman_source_term_device.f90 [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  use device_math, only : device_subcol3
+  use field_registry, only : neko_field_registry
+  implicit none
+  private
+
+  public :: brinkman_source_term_compute_device
+
+contains
+
+  !> Computs the constant source term on the cpu.
```
Comment (timofeymukha) [inline src/source_terms/bcknd/device/brinkman_source_term_device.f90:47]:
```suggestion
  !> Computes the Brinkman source term on the device.
```

File: src/source_terms/bcknd/cpu/filters_cpu.f90 [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+contains
+
+  ! ========================================================================== !
+  ! Internal functions and subroutines
+  ! ========================================================================== !
+
+  !> @brief Apply a smooth step function to a scalar.
+  elemental function smooth_step_cpu(x, edge0, edge1) result(res)
+    real(kind=rp), intent(in) :: x
```
Comment (timofeymukha) [inline src/source_terms/bcknd/cpu/filters_cpu.f90:46]: @njansson Are `elemental` ok? I remember you mentioning compiler issues.

File: examples/immersed_bunny/readme.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Immersed boundary of a Stanford Bunny
+
+This example illustrates the brinkman condition source term used to simulate an
+immersed boundary. A Stanford Buny model is placed in side a square duct and
```
Comment (timofeymukha) [inline examples/immersed_bunny/readme.md:4]:
```suggestion
immersed boundary. A Stanford Bunny model is placed in side a square duct and
```

File: examples/immersed_bunny/readme.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+Bunny url:              https://www.thingiverse.com/thing:151081
+Bunny domain:           [ [-24, 85], [-41, 45], [5, 112] ]
+Design domain:          [ [-200, 600], [-100, 100], [0, 200] ]
+```
+
+The run script will construct a design domain as above with 16 x 8 x 8 elements.
+The user can decide to specify another resolution from the terminal.
+
+The user can change how sharp the interface between fluid and solid is by
+changing the `distance_transform.value`, the value specify the width of the
```
Comment (timofeymukha) [inline examples/immersed_bunny/readme.md:17]:
```suggestion
changing the `distance_transform.value`, the value specifies the width of the
```

File: src/source_terms/brinkman_source_term.f90 [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+  use neko_config, only: NEKO_BCKND_DEVICE
+  use utils, only: neko_error
+  use brinkman_source_term_cpu, only: brinkman_source_term_compute_cpu
+  use brinkman_source_term_device, only: brinkman_source_term_compute_device
+  implicit none
+  private
+
+  !> A constant source term.
+  !! The strength is specified with the `values` keyword, which should be an
+  !! array, with a value for each component of the source.
```
Comment (timofeymukha) [inline src/source_terms/brinkman_source_term.f90:52]: Copy-pasted :-)

File: src/source_terms/brinkman_source_term.f90 [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    end select
+
+  end subroutine brinkman_source_term_init_from_json
+
+  !> Destructor.
+  subroutine brinkman_source_term_free(this)
+    class(brinkman_source_term_t), intent(inout) :: this
+
+    call this%free_base()
+  end subroutine brinkman_source_term_free
```
Comment (timofeymukha) [inline src/source_terms/brinkman_source_term.f90:135]: Nullify the brinkman pointer.

File: src/source_terms/brinkman/aabb.f90 [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+     procedure, pass(this), public :: min_distance => aabb_min_distance
+
+     ! Comparison operators
+     generic :: operator(.lt.) => less
+     generic :: operator(.gt.) => greater
+
+     procedure, pass(this), public :: overlaps => aabb_overlaps
+     procedure, pass(this), public :: contains => aabb_contains_other
+     procedure, pass(this), public :: contains_point => aabb_contains_point
+
```
Comment (timofeymukha) [inline src/source_terms/brinkman/aabb.f90:130]: Could you please copy-paste the "brief" part of the docstrings here as well?

File: src/source_terms/brinkman/aabb_tree.f90 [hunk 9]
```diff
@@ -0,0 +1,915 @@
...
+    type(aabb_node_t) :: leaf_node
+    type(aabb_node_t) :: tree_node
+    type(aabb_node_t) :: left_node
+    type(aabb_node_t) :: right_node
+
+    type(aabb_t) :: combinedaabb
+    real(kind=rp) :: newParentNodeCost
+    real(kind=rp) :: minimumPushDownCost
+    type(aabb_t) :: newLeftaabb
+    type(aabb_t) :: newRightaabb
```
Comment (timofeymukha) [inline src/source_terms/brinkman/aabb_tree.f90:668]: A little unorthodox with camelCase names in Neko.

**Review by timofeymukha — CHANGES_REQUESTED** [review]

**Comment by timofeymukha** [comment]
Not gonna lie, I did not check any of the algorithms in the tree and distance stuff. I guess in an ideal world, it would be nice to add unit tests for the aabb, the tree and the distance. Would double as a demonstrator for how these things work :). But perhaps in a later PR then, in case someone feels inspired.

Question, should the aabb go to "common"? Seems like a larger scope than just this source term, e.g. already discussed in context of probes.

**Comment by timfelle** [comment]
> Not gonna lie, I did not check any of the algorithms in the tree and distance stuff. I guess in an ideal world, it would be nice to add unit tests for the aabb, the tree and the distance. Would double as a demonstrator for how these things work :). But perhaps in a later PR then, in case someone feels inspired.
> 
> Question, should the aabb go to "common"? Seems like a larger scope than just this source term, e.g. already discussed in context of probes.

Yes, i would like to write some tests for it. However, i have had some issues setting up my environment with the test library. But i am almost there, i think..

And sure, lets move the general stuff somewhere else. I think the signed distance stuff could go to math and the aabb_tree could go to mesh, since thats where the octree also lives.

File: src/source_terms/brinkman/aabb_tree.f90 (hunk 9 above)
Comment (timfelle) [reply src/source_terms/brinkman/aabb_tree.f90:668]: Yea, not on purpose. It was a leftover from the original source code that i
overlooked when changing the others.

**Comment by timfelle** [comment]
@njansson Any idea why the MacOS one fails here?
It seem that when I re-ran it different tests from the pfunit thing fails.

Additionally the runner seem upset about some MacOS version stuff:
 was built for newer 'macOS' version (13.6) than being linked (13.0)

File: src/source_terms/bcknd/cpu/filters_cpu.f90 (hunk 3 above)
Comment (njansson) [reply src/source_terms/bcknd/cpu/filters_cpu.f90:46]: It's fine, but performance is a bit uncertain (if the compiler is bad...), so in short OK for this limited use

**Comment by njansson** [comment]
> @njansson Any idea why the MacOS one fails here? It seem that when I re-ran it different tests from the pfunit thing fails.
> 
> Additionally the runner seem upset about some MacOS version stuff: was built for newer 'macOS' version (13.6) than being linked (13.0)

There seems to be some issue with the runner, it fails when trying to create some temp storage (according to the failed test logs)

**Review by timofeymukha — APPROVED** [review]

**Pull request merged**